### PR TITLE
TeamCity: Add validating GHAs that check Kotlin files listing services

### PR DESCRIPTION
**Pull request by BBBmau**
Closes [#17168](https://github.com/hashicorp/terraform-provider-google/issues/17168)

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

**Comment by BBBmau** [comment]
Some notes:

it seemed to make sense to keep the `diff_check.go` file within the `.teamcity` folder however I understand that it could be moved to a more appropriate place within the `.teamcity` directory.

With the time it takes for the provider to be built and since the provider has to be built regardless for unit-tests, it would be good to have it use the built provider from the unit-tests for it to be used in the teamcity-diff-check. Although, since we don't expect new services to be added we could just ignore this also.

File: mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go:39]: Instead of collecting this information as a long slice of bytes, I'd recommend collecting a list of the service names as human readable strings. This would make the tool easier to debug during development and also help with making actionable error messages when a service is identified as missing.

The loop over submatches on lines 42-46 could contain something like:

```
dst := []byte{}
service := pattern.Expand(dst, template, stdout, submatches)
googleServices = append(googleServices, string(service))
```

---

Following that, another idea to think about might be storing the found services in a map where the key is the service name, instead of using a list. This would make it much easier to identify if a service is present or not:

```
// assuming data gets stored as map[string]bool (value in key-value pair not super important)

// from processing services_ga.kt
servicesInTeamCity := map[string]bool {
  "serviceA" : true,
  // serviceB isn't a key because it's missing from the services_ga.kt file used to make the map
  "serviceC" : true,
  ...
}

// from processing stdout of go list command
servicesInGaProvider := map[string]bool {
  "serviceA" : true,
  "serviceB" : true,
  "serviceC" : true,
  ...
}

for service, _ := range servicesInGaProvider {
  val, ok := servicesInTeamCity[service]; !ok {
    fmt.Printf("service %s is in the GA provider but not in services_gs.kt!", service)
  }
}
```

File: mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go:37]: Is the `\n` needed here?

File: mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go
Comment (SarahFrench) [reply mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go:24]: From taking a look at this code I think flags to receive the path to the service_x.kt Kotlin file and the path to the services folder in the unzipped artifact containing the TPG/TPGB codebase would be a good idea!

File: mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go:33]: If "google" / "google-beta" is obtained as an argument, or can be inferred from other arguments, that could be used to produce a slightly simpler regex.

**Review by SarahFrench — COMMENTED** [review]
Thanks for sharing this @BBBmau! It's a great start and I like how you've navigated pulling information out of the Kotlin map!

You asked about where to store this work/the diff_check.go file. I'd suggest moving the file into it's own folder in [the tools folder of magic modules](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/tools). This would mean that the GHA would need to both checkout the MM repo (to get access to the tool) and also open the TPG/TPGB codebase via the `Download built artifacts` step (to be able to run go list). This would also help you navigate [the problem in this test GHA run of yours](https://github.com/BBBmau/magic-modules/actions/runs/7925793270/job/21639590507), which fails because the google-beta provider doesn't include the .teamcity folder.

File: mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go
Comment (BBBmau) [reply mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go]: Removed since I found that it's really not necessary for template

File: tools/teamcity-diff-check/main.go
Comment (SarahFrench) [inline tools/teamcity-diff-check/main.go:67]: The confusion about the GHA working with GA but not Beta is because this GHA should validate the service files present in the magic-modules repo, not files copied into the downstream provider repos. I think this change is needed, but you'll need to test it:

```suggestion
	f, err := os.Open(fmt.Sprintf("../mmv1/third_party/terraform/.teamcity/components/inputs/%s", *serviceFile))
```

Only the GA version of the provider contains a .teamcity folder, which is why the GHA currently fails when it's attempting to check the services_beta.kt file in the Beta provider code base. Only the GA provider having a .teamcity folder is intentional.

The GHAs need to download the GA and Beta artefacts to perform the `go list` step in your script and determine which service packages are present, but otherwise you shouldn't need to interact with the downstream providers' files.

File: tools/teamcity-diff-check/main.go
Comment (SarahFrench) [inline tools/teamcity-diff-check/main.go:104]: Could you please implement a way of telling users which service(s) is missing? Currently, if this GHA identifies there's a diff the user will need to compare a list of ~135 folders to the map of services in the matching Kotlin file to find out what fix is required

File: tools/teamcity-diff-check/main.go
Comment (BBBmau) [reply tools/teamcity-diff-check/main.go]: This is true, the correct way where `services in the provider` is the ground truth would be to compare the services found in teamcity with that of google services (which would be a complete flip in how it's shown in the code).

File: tools/teamcity-diff-check/main.go
Comment (BBBmau) [reply tools/teamcity-diff-check/main.go]: https://github.com/BBBmau/magic-modules/actions/runs/7992740233/job/21827124224

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [inline .github/workflows/teamcity-services-diff-check.yml:85]: Looks like an accidental duplication here:

```suggestion
```

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [inline .github/workflows/teamcity-services-diff-check.yml]: Lets pair on this section, but here's a list of things that we should discuss when we pair:

- What is the if condition checking here? Is it correct?
- Adding feedback to the user about the significance of this step, e.g. via an `echo` command
- Ways to make the newServices count more accurate using patterns/regex with `grep`
   -  On Mac, `grep -e "mmv1/products/.*/product.yaml"` would help identify when new services are added
- Use of [outputs](https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs) and [conditional job execution](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution) to control when the GHA runs (replacing `exit 1`)

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [inline .github/workflows/teamcity-services-diff-check.yml:44]:
```suggestion
      - name: Download built artifacts - GA provider
```

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [inline .github/workflows/teamcity-services-diff-check.yml:55]:
```suggestion
      - name: Download built artifacts - Beta provider
```

**Review by SarahFrench — COMMENTED** [review]
Thanks for taking on my feedback! I think we should pair on making this GHA only run fully if new services are added, but in my review comments I left a bullet point list of things we should discuss while pairing. I'll put something in our calendars

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [inline .github/workflows/teamcity-services-diff-check.yml:11]: I'm glad we established that the use of paths here + the `git diff` step will help us identify a new service being added pretty accurately. Though after thinking that through I realised that another edge case is probably when a service is promoted to GA. In that situation services_ga.kt will be mis-matched with the GA provider but this GHA won't have an opportunity to be triggered automatically.

I think a solution to this is adding a [schedule](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) trigger and making it run once a week. [This is a useful website](https://crontab.guru/) to check the cron expression. I'd recommend running this on early Tuesday mornings (i.e. sometime ~4am UTC) because that's the last time we could catch a newly-added service missing from TeamCity before the the tests run on Wednesday morning and are used for the release cut.

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [inline .github/workflows/teamcity-services-diff-check.yml:22]: Could you add a `name` property here to help indicate what the step is doing, e.g. "Check for new services"

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [reply .github/workflows/teamcity-services-diff-check.yml:11]: Given the PR-specific stuff below using `git diff`, it might be that we can't add a schedule trigger. Instead we might need to pull out some of the stuff below into reusable workflows, and then create a PR-triggered and schedule-triggered version of the workflow that both use the TeamCity-checking steps.

**Review by SarahFrench — COMMENTED** [review]
Thanks for bearing with me - here are some more thoughts on when to trigger the GHA

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [reply .github/workflows/teamcity-services-diff-check.yml]: Addressed by adding explicit flags for both GA and Beta

File: .github/workflows/teamcity-services-diff-check.yml
Comment (SarahFrench) [reply .github/workflows/teamcity-services-diff-check.yml:11]: See also: https://github.com/hashicorp/terraform-provider-google/issues/15633#issuecomment-1961697734

File: tools/teamcity-diff-check/main.go
Comment (BBBmau) [reply tools/teamcity-diff-check/main.go:48]: This would be good to follow, this would also require another flag to include the use of the `service_file.txt` which is where all services are stored from the `git diff` command. This is due to the `service_file` flag also being used for the `.txt` file also

They could be separated to be two different flags: `-servicekt_path` and `-service_txt`

For now we can save this for another time 👍🏼

File: .github/workflows/teamcity-services-diff-check.yml
Comment (BBBmau) [reply .github/workflows/teamcity-services-diff-check.yml:11]: Latest commit adds an almost identical workflow that runs on PRs, only difference is the removal of `check-pr` job and the addition of `schedule` where it'll run every tuesday morning (4am UTC)

Comment (SarahFrench) [reply .github/workflows/teamcity-services-diff-check.yml:11]: Looks good! A future improvement could be to use reusable workflows to reduce repetition, but lets leave that for now and get this merged 🚢

**Review by SarahFrench — APPROVED** [review]
LGTM - once this is merged lets open some pretend PRs to confirm that it's behaving as expected, just to be certain

**Pull request merged**
